### PR TITLE
Clean Unit test up (favor ExpectedException over try catch fail)

### PR DESCRIPTION
**Pull request by vanniktech**

### CODE REVIEW

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 1]
```diff
@@ -1,75 +1,55 @@
...
 
+import org.junit.Rule;
 import org.junit.Test;
-
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.fail;
+import org.junit.rules.ExpectedException;
 
 public class PrivateConstructionCheckerTest {
+    @Rule public ExpectedException expectedException = ExpectedException.none();
```
Comment (nikitin-da) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java:8]: Good approach

**Comment by nikitin-da** [comment]
LGTM @artem-zinnatullin  PTAL, please

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 2]
```diff
@@ -1,75 +1,55 @@
...
-
-            fail();
-        } catch (IllegalArgumentException expected) {
-            assertEquals("expectedTypeOfException can not be null", expected.getMessage());
-        }
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("expectedTypeOfException can not be null");
+        PrivateConstructorChecker
+                .forClass(Object.class)
+                .expectedTypeOfException(null);
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Can you please divide `forClass` and `expectedTypeOfException`, I mean, to make sure that it fails after call to the `expectedTypeOfException` and not to the `forClass`

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 3]
```diff
@@ -1,75 +1,55 @@
...
-
-            fail();
-        } catch (IllegalArgumentException expected) {
-            assertEquals("expectedExceptionMessage can not be null", expected.getMessage());
-        }
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("expectedExceptionMessage can not be null");
+        PrivateConstructorChecker
+                .forClass(Object.class)
+                .expectedExceptionMessage(null);
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 4]
```diff
@@ -1,75 +1,55 @@
...
-
-            fail();
-        } catch (IllegalArgumentException expected) {
-            assertEquals("class can not be null", expected.getMessage());
-        }
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("class can not be null");
+        PrivateConstructorChecker
+                .forClass(null)
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 5]
```diff
@@ -1,75 +1,55 @@
...
-
-            fail();
-        } catch (IllegalArgumentException expected) {
-            assertEquals("classes can not be null or empty", expected.getMessage());
-        }
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("classes can not be null or empty");
+        PrivateConstructorChecker
+                .forClasses(null)
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 6]
```diff
@@ -1,75 +1,55 @@
...
-
-            fail();
-        } catch (IllegalArgumentException expected) {
-            assertEquals("class can not be null", expected.getMessage());
-        }
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("class can not be null");
+        PrivateConstructorChecker
+                .forClasses(ClassWithPrivateConstructor.class, null)
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 7]
```diff
@@ -79,18 +59,11 @@ private ClassWithoutDefaultConstructor(String someParam) {
...
-            assertEquals(
-                    ClassWithoutDefaultConstructor.class + " has non-default constructor with some parameters",
-                    expected.getMessage()
-            );
-        }
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage(ClassWithoutDefaultConstructor.class + " has non-default constructor with some parameters");
+        PrivateConstructorChecker
+                .forClass(ClassWithoutDefaultConstructor.class)
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 8]
```diff
@@ -112,15 +85,11 @@ public void shouldAssertThatConstructorIsPrivateAndDoesNotThrowExceptions() {
...
-
-            fail();
-        } catch (AssertionError expected) {
-            assertEquals("Constructor of " + ClassWithDefaultProtectedConstructor.class + " must be private", expected.getMessage());
-        }
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage("Constructor of " + ClassWithDefaultProtectedConstructor.class + " must be private");
+        PrivateConstructorChecker
+                .forClass(ClassWithDefaultProtectedConstructor.class)
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 9]
```diff
@@ -130,15 +99,11 @@ protected ClassWithProtectedConstructor() {
...
-
-            fail();
-        } catch (AssertionError expected) {
-            assertEquals("Constructor of " + ClassWithProtectedConstructor.class + " must be private", expected.getMessage());
-        }
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage("Constructor of " + ClassWithProtectedConstructor.class + " must be private");
+        PrivateConstructorChecker
+                .forClass(ClassWithProtectedConstructor.class)
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 10]
```diff
@@ -148,15 +113,11 @@ public ClassWithPublicConstructor() {
...
-
-            fail();
-        } catch (AssertionError expected) {
-            assertEquals("Constructor of " + ClassWithPublicConstructor.class + " must be private", expected.getMessage());
-        }
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage("Constructor of " + ClassWithPublicConstructor.class + " must be private");
+        PrivateConstructorChecker
+                .forClass(ClassWithPublicConstructor.class)
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 11]
```diff
@@ -192,50 +153,36 @@ public void shouldCheckThatConstructorThrowsExceptionWithExpectedMessageButWitho
...
-            );
-        }
+        expectedException.expect(IllegalStateException.class);
+        expectedException.expectMessage("For " + ClassWithConstructorThatThrowsException.class + " expected exception of type = class java.lang.IllegalArgumentException, " +
+                "but was exception of type = class java.lang.IllegalStateException");
+
+        PrivateConstructorChecker
+                .forClass(ClassWithConstructorThatThrowsException.class)
+                .expectedTypeOfException(IllegalArgumentException.class) // Incorrect type
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 12]
```diff
@@ -192,50 +153,36 @@ public void shouldCheckThatConstructorThrowsExceptionWithExpectedMessageButWitho
...
-        }
+        expectedException.expect(IllegalStateException.class);
+        expectedException.expectMessage("For " + ClassWithConstructorThatThrowsException.class + " expected exception message = 'lol, not something that you've expected?', " +
+                "but was = 'test exception'");
+
+        PrivateConstructorChecker
+                .forClass(ClassWithConstructorThatThrowsException.class)
+                .expectedTypeOfException(IllegalStateException.class) // Correct type
+                .expectedExceptionMessage("lol, not something that you've expected?") // Incorrect message
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 13]
```diff
@@ -192,50 +153,36 @@ public void shouldCheckThatConstructorThrowsExceptionWithExpectedMessageButWitho
...
-
-            fail();
-        } catch (IllegalStateException expected) {
-            assertEquals("For " + ClassWithConstructorThatThrowsException.class + " no exception was expected", expected.getMessage());
-        }
+        expectedException.expect(IllegalStateException.class);
+        expectedException.expectMessage("For " + ClassWithConstructorThatThrowsException.class + " no exception was expected");
+        PrivateConstructorChecker
+                .forClass(ClassWithConstructorThatThrowsException.class)
+                .check(); // We don't expect exception, but it will be thrown
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 14]
```diff
@@ -252,15 +199,11 @@ private ClassWith2Constructors(String str) {
...
-
-            fail();
-        } catch (AssertionError expected) {
-            assertEquals(ClassWith2Constructors.class + " has more than one constructor", expected.getMessage());
-        }
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage(ClassWith2Constructors.class + " has more than one constructor");
+        PrivateConstructorChecker
+                .forClass(ClassWith2Constructors.class)
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 15]
```diff
@@ -269,15 +212,11 @@ public void shouldThrowExceptionBecauseClassHasMoreThanOneConstructor() {
...
-
-            fail();
-        } catch (AssertionError expected) {
-            assertEquals("Constructor of " + ClassWithoutDeclaredConstructor.class + " must be private", expected.getMessage());
-        }
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage("Constructor of " + ClassWithoutDeclaredConstructor.class + " must be private");
+        PrivateConstructorChecker
+                .forClass(ClassWithoutDeclaredConstructor.class)
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

File: checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java [hunk 16]
```diff
@@ -297,17 +236,10 @@ public void shouldCheckMultipleThatConstructorThrowsExceptionWithExpectedMessage
...
-            assertEquals(
-                    ClassWithoutDefaultConstructor.class + " has non-default constructor with some parameters",
-                    expected.getMessage()
-            );
-        }
+        expectedException.expect(AssertionError.class);
+        expectedException.expectMessage(ClassWithoutDefaultConstructor.class + " has non-default constructor with some parameters");
+        PrivateConstructorChecker
+                .forClasses(ClassWithPrivateConstructor.class, ClassWithoutDefaultConstructor.class)
+                .check();
```
Comment (artem-zinnatullin) [inline checker/src/test/java/com/pushtorefresh/private_constructor_checker/PrivateConstructionCheckerTest.java]: Same as above

**Comment by artem-zinnatullin** [comment]
Sorry for tons of same comments, but they'll help during next iteration of code review

**Comment by vanniktech** [comment]
@artem-zinnatullin don't worry. You're right. I hope I got every place.

**Comment by artem-zinnatullin** [comment]
:+1:

**Comment by artem-zinnatullin** [comment]
@vanniktech thanks!

**Pull request merged**